networkx/algorithms/community/louvain.py

```diff
@@ -0,0 +1,531 @@
+    best_partition
+
+    Notes
+    -----
+    Uses Louvain algorithm
+
+    References
+    ----------
+    .. 1. Blondel, V.D. et al. Fast unfolding of communities in large networks. J. Stat. Mech 10008, 1-12(2008).
+
+    Examples
+    --------
+    >>> G=nx.erdos_renyi_graph(100, 0.01)
+    >>> dendo = generate_dendogram(G)
+    >>> for level in range(len(dendo) - 1) :
+    >>>     print "partition at level", level, "is", partition_at_level(dendo, level)
+    """
+    if type(graph) != nx.Graph :
+        raise TypeError("Bad graph type, use only non directed graph")
+
+    #special case, when there is no link
+    #the best partition is everyone in its community
+    if graph.number_of_edges() == 0 :
+        part = dict([])
+        for node in graph.nodes() :
+            part[node] = node
+        return part
+
+    current_graph = graph.copy()
+    status = Status()
+    status.init(current_graph, part_init)
+    mod = __modularity(status)
+    status_list = list()
+    __one_level(current_graph, status)
+    new_mod = __modularity(status)
+    partition = __renumber(status.node2com)
+    status_list.append(partition)
+    mod = new_mod
+    current_graph = induced_graph(partition, current_graph)
+    status.init(current_graph)
+
+    while True :
+        __one_level(current_graph, status)
+        new_mod = __modularity(status)
+        if new_mod - mod < __MIN :
+            break
+        partition = __renumber(status.node2com)
+        status_list.append(partition)
+        mod = new_mod
+        current_graph = induced_graph(partition, current_graph)
+        status.init(current_graph)
+    return status_list[:]
+
+
+def induced_graph(partition, graph) :
+    """Produce the graph where nodes are the communities
+
+    there is a link of weight w between communities if the sum of the weights of the links between their elements is w
+
+    Parameters
+    ----------
+    partition : dict
+       a dictionary where keys are graph nodes and  values the part the node belongs to
+    graph : networkx.Graph
+        the initial graph
+
+    Returns
+    -------
+    g : networkx.Graph
+       a networkx graph where nodes are the parts
+
+    Examples
+    --------
+    >>> n = 5
+    >>> g = nx.complete_graph(2*n)
+    >>> part = dict([])
+    >>> for node in g.nodes() :
+    >>>     part[node] = node % 2
+    >>> ind = induced_graph(part, g)
+    >>> goal = nx.Graph()
+    >>> goal.add_weighted_edges_from([(0,1,n*n),(0,0,n*(n-1)/2), (1, 1, n*(n-1)/2)])
+    >>> nx.is_isomorphic(int, goal)
+    True
+    """
+    ret = nx.Graph()
+    ret.add_nodes_from(partition.values())
+
+    for node1, node2, datas in graph.edges_iter(data = True) :
+        weight = datas.get("weight", 1)
+        com1 = partition[node1]
+        com2 = partition[node2]
+        w_prec = ret.get_edge_data(com1, com2, {"weight":0}).get("weight", 1)
+        ret.add_edge(com1, com2, weight = w_prec + weight)
+
+    return ret
+
+
+def __renumber(dictionary) :
```

Comment:
Better:

    vals = set(d.values())
    mapping = {val: i for i, val in enumerate(vals)}
    return {k: mapping[v] for k, v in d.items()}